log-manager/src/main/java/io/airlift/log/RollingFileHandler.java

```diff
@@ -92,74 +234,352 @@ public void publish(LogRecord record)
         }
 
         try {
-            asyncAppender.doAppend(message);
+            putUninterruptibly(queue, message);
         }
         catch (Exception e) {
             // catch any exception to assure logging always works
             reportError(null, e, WRITE_FAILURE);
         }
+
+        // if closed while queueing, try to remove the message just to clean up
+        if (closed.get()) {
+            queue.remove(message);
+        }
     }
 
     @Override
-    public void flush()
+    public synchronized void flush()
     {
+        if (currentOutputStream != null) {
+            try {
+                currentOutputStream.flush();
+            }
+            catch (Exception e) {
+                reportError(null, e, FLUSH_FAILURE);
+            }
+        }
     }
 
     @Override
     public void close()
+            throws SecurityException
     {
+        closed.set(true);
+
+        putUninterruptibly(queue, POISON_MESSAGE);
+
+        // wait for logging to finish
         try {
-            asyncAppender.stop();
+            thread.join();
         }
-        catch (Exception e) {
-            // catch any exception to assure logging always works
-            reportError(null, e, CLOSE_FAILURE);
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
```

Comment:
this is erroneous situation and not handled correctly by the following code, so
log (stderr) that things are going wrong

(or add a comment why not doing so)

Reply:
same for the other catch below

Reply:
The rules for interruption is you should (with very few exception), throw an InterruptedException or set the interrupted flag on the thread.

The code here attempts to do a clean shutdown unless the interrupted flag is set, and in that case we cleanup as quickly as possible.